module/input/SpeechIntent/src/SpeechIntent.cpp

```diff
@@ -0,0 +1,391 @@
+            return;
+        }
+
+        //set "debug_mode: true" in SpeechIntent.yaml to enable this
+        if (this->config.debug_mode)
+        {
+            on<Trigger<SpeechIntentMessage>>().then([this](const SpeechIntentMessage& msg) {
+                log(fmt::format("text = {} ({})", msg.text.data(), msg.text.data()));
+                log(fmt::format("intent = {} ({})", msg.intent.data(), msg.intent.data()));
+                for (Slot slot : msg.slots) {
+                    log(fmt::format("{} = {}", slot.name.data(), slot.value.data()));
+                }
+            });
+            // this is just for testing
+            // we need this as we do not have an
+            // environment for testing the microphone
+            // allows us to simulate sending a MicControlMsg
+            // as if we were another module
+            on<IO>(STDIN_FILENO, IO::READ).then([this] {
+                std::string str = {};
+                getline(std::cin, str);
+
+                log(fmt::format("SpeechIntent STDIN = {}", str));
+
+                std::stringstream ss(str);
+                std::string cmd = {};
+                getline(ss, cmd, ' ');
+                if (cmd == "file") {
+                    std::string filename = {};
+                    getline(ss, filename);
+                    std::unique_ptr<MicControlMsg> msg =
+                        std::make_unique<MicControlMsg>(MicControlMsg{MIC_MSG_RECOGNIZE_AUDIO_FROM_FILE, filename});
+                    emit(std::move(msg));
+                }
+            });
+        }
+
+        on<Trigger<MicControlMsg>>().then([this](const MicControlMsg& msg) {
+            // Decide on appropriate action based on msg type
+            switch (msg.type) {
+                case MIC_MSG_ENABLE: enabled = true; break;
+                case MIC_MSG_DISABLE: enabled = false; break;
+                case MIC_MSG_RECOGNIZE_AUDIO_FROM_FILE:
+                    if (!write_audio_to_file(voice2json_proc.stdin, (char*) msg.filename.data())) {
+                        NUClear::log<NUClear::FATAL>(
+                            fmt::format("({}:{}) Failed to write to {}", __FILE__, __LINE__, msg.filename.data()));
+                    }
+                    break;
+            }
+        });
+
+
+        on<IO>(voice2json_proc.stdout, IO::READ).then([this] {
+            // TODO: need to read and append to buffer until we reach a newline...
+            char buffer[0x1000];
+            ssize_t bytes_read = read(voice2json_proc.stdout, buffer, sizeof(buffer) - 1);
+            if (bytes_read == -1) {
+                NUClear::log<NUClear::FATAL>(
+                    fmt::format("({}:{}) failed to read from spawned process stdout, errno = {}",
+                                __FILE__,
+                                __LINE__,
+                                errno));
+                return;
+            }
+            buffer[bytes_read] = 0;
+
+            if (enabled) {
```

Comment:
What is enabled?

Reply:
That is a bool that controls the behaviour of whether or not the module should respond to input reading from voice2json. This is useful when running in "transcribe-stream" mode (continuously running). When enabled the module will attempt to parse the voice2json json and emit it as a SpeechIntentMessage. When disabled it won't emit anything. Perhaps it deserves a more descriptive name.

Reply:
Ok, it's been renamed to "output_enabled"

Reply:
I'd have to read the code closer, but why run the reaction if the pasing is not enabled? You can just disable the reaction

Reply:
For the moment we are contiguously running the voice2json process in either transcribe-stream or transcribe-wav mode. This implies that we have to read from the voice2sjon stdout pipe when data is available. If we fail to read from it then the pipe will continue to fill up indefinitely. We may choose to change this in the future so that you can selectively turn voice2json on or off.